### ISSUE TITLE
Unable to install Nvidia GPU driver in TDX kernel

### ISSUE THREAD

**Issue by iihihiuh** [intel/tdx-tools#540]
After launching a TDX VM (tag 2023ww15 `start-qemu.sh`) , I was trying to install the GPU [driver](https://developer.download.nvidia.com/compute/cuda/12.2.1/local_installers/cuda_12.2.1_535.86.10_linux.run) inside TDX VM (Nvidia H100 with CC on/off). However, the installation was not successful producing error message below:

```
[  187.666645] nvidia: loading out-of-tree module taints kernel.
[  187.700935] nvidia-nvlink: Nvlink Core is being initialized, major device number 237
[  187.700940] NVRM: The NVIDIA probe routine was not called for 1 device(s).
[  187.701537] NVRM: This can occur when a driver such as: 
               NVRM: nouveau, rivafb, nvidiafb or rivatv 
               NVRM: was loaded and obtained ownership of the NVIDIA device(s).
[  187.701539] NVRM: Try unloading the conflicting kernel module (and/or
               NVRM: reconfigure your kernel without the conflicting
               NVRM: driver(s)), then try loading the NVIDIA kernel module
               NVRM: again.
[  187.701540] NVRM: No NVIDIA devices probed.
[  187.701724] nvidia-nvlink: Unregistered Nvlink Core, major device number 237
```

Although when booting the TDX VM with PCI passthrough, qemu produces an error saying "VFIO_MAP_DMA failed: No space left on device", I can still see my GPU using `lspci`.

Does anyone have some ideas on how to resolve this issue?

**Comment by zvonkok** [comment]
```
 NVRM: This can occur when a driver such as: 
               NVRM: nouveau, rivafb, nvidiafb or rivatv 
```

Did you blacklist nouveau?

**Comment by iihihiuh** [comment]
I did not blacklist nouveau, because I did not see any of those drivers when running `lsmod`.

**Comment by iihihiuh** [comment]
> ```
>  NVRM: This can occur when a driver such as: 
>                NVRM: nouveau, rivafb, nvidiafb or rivatv 
> ```
> 
> Did you blacklist nouveau?
Thank you for your quick reply.

Actually, yes. I blacked-list nouveau, and this issue persists.

**Comment by zvonkok** [comment]
Can you do a `lspci -s <BDF OF GPU> -vvv`

**Comment by zvonkok** [comment]
```
❯ lspci | grep NVIDIA 
01:00.0 VGA compatible controller: NVIDIA Corporation GA104GLM [RTX A3000 Mobile] (rev a1)
01:00.1 Audio device: NVIDIA Corporation GA104 High Definition Audio Controller (rev a1)
````

**Comment by iihihiuh** [comment]
> ```
> ❯ lspci | grep NVIDIA 
> 01:00.0 VGA compatible controller: NVIDIA Corporation GA104GLM [RTX A3000 Mobile] (rev a1)
> ^BDF^
> ```

Below is the lspci output. Please let me know if you need further information.
```
01:00.0 3D controller: NVIDIA Corporation Device 2331 (rev a1)
        Subsystem: NVIDIA Corporation Device 1626
        Physical Slot: 0
        Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 32 bytes
        Interrupt: pin A routed to IRQ 11
        Region 0: Memory at 702002000000 (64-bit, prefetchable) [size=16M]
        Region 2: Memory at 700000000000 (64-bit, prefetchable) [size=128G]
        Region 4: Memory at 702000000000 (64-bit, prefetchable) [size=32M]
        Capabilities: [40] Power Management version 3
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=0mA PME(D0-,D1-,D2-,D3hot-,D3cold-)
                Status: D0 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [48] MSI: Enable- Count=1/16 Maskable+ 64bit+
                Address: 0000000000000000  Data: 0000
                Masking: 00000000  Pending: 00000000
        Capabilities: [60] Express (v2) Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0, Latency L0s <64ns, L1 <1us
                        ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset+ SlotPowerLimit 25.000W
                DevCtl: CorrErr- NonFatalErr- FatalErr+ UnsupReq-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+ FLReset-
                        MaxPayload 128 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr- TransPend-
                LnkCap: Port #0, Speed 32GT/s, Width x16, ASPM L1, Exit Latency L1 <4us
                        ClockPM+ Surprise- LLActRep- BwNot- ASPMOptComp+
                LnkCtl: ASPM Disabled; RCB 64 bytes, Disabled- CommClk-
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 32GT/s (ok), Width x16 (ok)
                        TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
                DevCap2: Completion Timeout: Range AB, TimeoutDis+ NROPrPrP- LTR+
                         10BitTagComp+ 10BitTagReq+ OBFF Via message, ExtFmt- EETLPPrefix-
                         EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
                         FRS- TPHComp- ExtTPHComp-
                         AtomicOpsCap: 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis- LTR- OBFF Disabled,
                         AtomicOpsCtl: ReqEn+
                LnkCap2: Supported Link Speeds: 2.5-32GT/s, Crosslink- Retimer+ 2Retimers+ DRS-
                LnkCtl2: Target Link Speed: 32GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete+ EqualizationPhase1+
                         EqualizationPhase2+ EqualizationPhase3+ LinkEqualizationRequest-
                         Retimer- 2Retimers- CrosslinkRes: unsupported
        Capabilities: [9c] Vendor Specific Information: Len=14 <?>
        Capabilities: [b0] MSI-X: Enable- Count=9 Masked-
                Vector table: BAR=0 offset=00b90000
                PBA: BAR=0 offset=00ba0000
        Kernel modules: nvidiafb, nvidia_drm, nvidia
```

**Comment by kpadwal** [comment]
To authorize the GPUs in the guest, update the guest command line.
authorize_allow_devs=pci:0x10de:2331

**Comment by iihihiuh** [comment]
> To authorize the GPUs in the guest, update the guest command line. authorize_allow_devs=pci:0x10de:2331

Thank u for ur suggestions. I tried as you suggested. During the driver installation the guest VM faulted and gave this error message:
```
qemu-system-x86_64: warning: Unknown start 0x702002000000 size 0x1000 shared_to_private 1
EAX=00000000 EBX=00000000 ECX=00000000 EDX=000c06f2
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000000
Code=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <00> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

It looks like the kernel completely faulted.

**Comment by ruomengh** [comment]
Thanks for raising the issue. The project is end of life. Please refer to Red Hat blog [Enabling hardware-backed confidential computing with a CentOS SIG](https://www.redhat.com/en/blog/enabling-hardware-backed-confidential-computing-centos-sig) or Ubuntu blog [Intel® TDX 1.0 technology preview available on Ubuntu 23.10](https://ubuntu.com/blog/intel-tdx-1-0-preview-on-ubuntu-23-10) or SUSE blog [Intel® TDX Support Coming to SUSE Linux Enterprise Server](https://www.suse.com/c/intel-tdx-support-coming-to-suse-linux-enterprise-server/) for more information of TDX Early Preview.
